include/power_grid_model/math_solver/math_solver.hpp

```diff
@@ -77,19 +106,39 @@ class MathSolver {
     void clear_solver() {
         newton_pf_solver_.reset();
         linear_pf_solver_.reset();
+        iterative_current_pf_solver_.reset();
+        linear_current_pf_solver_.reset();
     }
 
     void update_value(std::shared_ptr<MathModelParam<sym> const> const& math_model_param) {
         y_bus_.update_admittance(math_model_param);
+        // Invalidate prefactorization of iterative/linear current solvers
+        if (iterative_current_pf_solver_.has_value()) {
+            iterative_current_pf_solver_.value().reset_lhs();
+        }
+        if (linear_current_pf_solver_.has_value()) {
+            linear_current_pf_solver_.value().reset_lhs();
+        }
+    }
+
+    template <typename T>
+    MathOutput<sym> run_power_flow_iterative(T& base_solver, PowerFlowInput<sym> const& input, double err_tol,
+                                             Idx max_iter, CalculationInfo& calculation_info) {
+        // Calling function of CRTP implemented iterative solvers
+        return base_solver.run_power_flow(y_bus_, input, err_tol, max_iter, calculation_info);
     }
 
    private:
     std::shared_ptr<MathModelTopology const> topo_ptr_;
     YBus<sym> y_bus_;
     bool all_const_y_;  // if all the load_gen is const element_admittance (impedance) type
+    bool all_const_i_;  // if all the load_gen is const current type
     std::optional<NewtonRaphsonPFSolver<sym>> newton_pf_solver_;
     std::optional<LinearPFSolver<sym>> linear_pf_solver_;
     std::optional<IterativeLinearSESolver<sym>> iterative_linear_se_solver_;
+    std::optional<IterativecurrentPFSolver<sym>> iterative_current_pf_solver_;
+    // Linear current is single iteration of iterative current
+    std::optional<IterativecurrentPFSolver<sym>> linear_current_pf_solver_;
```

Comment:
Why do you need a separate solver instance for linear current? You can just call the same iterative current solver with one iteration and relaxed error tolerance.